src/playwright/matchers/to-dispatch-events.ts

```diff
@@ -12,115 +12,148 @@ interface SerializableEvent {
 export default expect.extend({
   async toDispatchEvents(
     locator: Locator,
-    action: () => Promise<unknown>,
-    expectedEvents: {
-      bubbles?: boolean;
-      cancelable?: boolean;
-      composed?: boolean;
-      defaultPrevented?: boolean;
-      type: string;
-    }[],
+    action: () => Promise<void>,
+    expectedEvents: Omit<SerializableEvent, 'timeStamp'>[],
   ) {
-    const receivedEventsPromise = locator.evaluate(async (element, events) => {
-      const promises = events.map(async (event) => {
-        const events: SerializableEvent[] = [];
-
-        await new Promise<void>((resolve, reject) => {
-          element.addEventListener(event.type, (event: Event) => {
-            events.push({
-              bubbles: event.bubbles,
-              cancelable: event.cancelable,
-              composed: event.composed,
-              defaultPrevented: event.defaultPrevented,
-              timeStamp: event.timeStamp,
-              type: event.type,
-            });
-
-            resolve();
-          });
-
-          setTimeout(() => {
-            // No need to reject with an `Error` given the rejection value is unused.
-            //
-            // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
-            reject(null);
+    await locator.evaluate((element, expectedEvents) => {
+      // Sadly, there doesn't appear to be another way. We need to be certain that the
+      // event listeners are in place before the action call is awaited. And the only
+      // way to ensure that is for the matcher to first await this evaluation.
+      //
+      // Before you scoff, as I did, at adding a property to the window, consider how
+      // useful this matcher is for testers. And consider whether this window nonsense
+      // will be flaky or unreliable in any way.
```

Comment:
Awaiting the evaluation means it has to resolve immediately after the listeners have been added (and not after events have dispatched). 

But then how do we know which events have been dispatched? Well, we have the listener add them to the window. Then (way below) we await the action. Then we run another evaluation to get the events from the window.